core/src/upgrade/transfer.rs

```diff
@@ -0,0 +1,436 @@
+    inner: WriteOneInner<TSocket, TData>,
+}
+
+enum WriteOneInner<TSocket, TData> {
+    /// We need to write the data length to the socket.
+    WriteLen(io::WriteAll<TSocket, io::Window<[u8; 10]>>, TData),
+    /// We need to write the actual data to the socket.
+    Write(io::WriteAll<TSocket, TData>),
+    /// We need to shut down the socket.
+    Shutdown(io::Shutdown<TSocket>),
+    /// A problem happened during the processing.
+    Poisoned,
+}
+
+impl<TSocket, TData> Future for WriteOne<TSocket, TData>
+where
+    TSocket: AsyncWrite,
+    TData: AsRef<[u8]>,
+{
+    type Item = ();
+    type Error = std::io::Error;
+
+    #[inline]
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        Ok(self.inner.poll()?.map(|_socket| ()))
+    }
+}
+
+impl<TSocket, TData> Future for WriteOneInner<TSocket, TData>
+where
+    TSocket: AsyncWrite,
+    TData: AsRef<[u8]>,
+{
+    type Item = TSocket;
+    type Error = std::io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            match mem::replace(self, WriteOneInner::Poisoned) {
+                WriteOneInner::WriteLen(mut inner, data) => match inner.poll()? {
+                    Async::Ready((socket, _)) => {
+                        *self = WriteOneInner::Write(io::write_all(socket, data));
+                    }
+                    Async::NotReady => {
+                        *self = WriteOneInner::WriteLen(inner, data);
+                    }
+                },
+                WriteOneInner::Write(mut inner) => match inner.poll()? {
+                    Async::Ready((socket, _)) => {
+                        *self = WriteOneInner::Shutdown(tokio_io::io::shutdown(socket));
+                    }
+                    Async::NotReady => {
+                        *self = WriteOneInner::Write(inner);
+                    }
+                },
+                WriteOneInner::Shutdown(ref mut inner) => {
+                    let socket = try_ready!(inner.poll());
+                    return Ok(Async::Ready(socket));
+                }
+                WriteOneInner::Poisoned => panic!(),
+            }
+        }
+    }
+}
+
+/// Reads a message from the given socket, then applies a transformation on it. Only one message
+/// is processed and the socket is dropped, because we assume that the socket will not send
+/// anything more.
+///
+/// The `max_size` parameter is the maximum size in bytes of the message that we accept. This is
+/// necessary in order to avoid DoS attacks where the remote sends us a message of several
+/// gigabytes.
+///
+/// > **Note**: Assumes that a variable-length prefix indicates the length of the message. This is
+/// >           compatible with what `write_one` does.
+#[inline]
+pub fn read_one<TSocket, TThen, TOut, TErr>(
+    socket: TSocket,
+    max_size: usize,
+    then: TThen,
+) -> ReadOne<TSocket, TThen>
+where
+    TThen: FnOnce(Vec<u8>) -> Result<TOut, TErr>,
```

Comment:
Instead of taking a `TThen` parameter, why could  the `ReadOne` future not just yield the `Vec<u8>` and the existing future combinators are used?

The Floodsub protocol implementation would be

```rust
upgrade::read_one(socket, 2048).from_err().and_then(|packet| { ... })
```

instead of

```rust
 upgrade::read_one(socket, 2048, |packet| { ... })
```

Granted, the associated `FloodsubConfig::Future` type is somewhat more difficult with

```rust
 type Future =
    AndThen<
        FromErr<upgrade::ReadOne<TSocket>, FloodsubDecodeError>,
        Result<FloodsubRpc, FloodsubDecodeError>,
        fn(Vec<u8>) -> Result<FloodsubRpc, FloodsubDecodeError>
    >;
```

But that just reflects the state of the current futures system and if performance is less of an issue one could always use trait objects instead.

Reply:
I'll add both versions.